cpp/src/io/utilities/memory_utils.hpp

```diff
@@ -0,0 +1,71 @@
+/*
+ * Copyright (c) 2018, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+	 * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/**
+ * @file memory_utils.hpp  code to provide RAII abstractions for RMM managed device memory
+ *
+ */
+
+#pragma once
+
+#include <type_traits>
+
+#include "rmm/rmm.h"
+#include "rmm/thrust_rmm_allocator.h"
+
+#include "utilities/error_utils.hpp"
+
+namespace rmm {
+
+// alias for the rmm unique pointer
+template<typename T>
+using unique_ptr = std::unique_ptr<T, std::function<void(std::remove_extent_t<T> *)>>;
+
+// creates a unique pointer holding a single object
+template<typename T, std::enable_if_t<!std::is_array<T>::value, int> = 0>
+unique_ptr<T> make_unique(cudaStream_t stream) {
+	T* ptr = nullptr;
+	const auto error = RMM_ALLOC(&ptr, sizeof(T), 0);
+	if(error != RMM_SUCCESS) {
+		cudf::detail::throw_cuda_error(cudaErrorMemoryAllocation, __FILE__, __LINE__);
+	}
+
+	auto deleter = [stream](T *p) {
+		RMM_FREE(p, stream);
+	};
+
+	return { ptr, deleter };
+}
+
+// creates a unique pointer holding an array of objects
+template<typename T, std::enable_if_t<std::is_array<T>::value, int> = 0>
+unique_ptr<T> make_unique(size_t size, cudaStream_t stream) {
```

Comment:
I like what you're ultimately trying to accomplish here, which is using RAII to manage your allocs/frees, however, I believe `unique_ptr` is a more specialized tool than what is necessary here. 

Instead, I believe the same goal can be accomplished more generally via a `rmm::device_buffer` class whose constructor calls `RMM_ALLOC` and destructor calls `RMM_FREE`. 

This will provide the RAII nice-ness for memory management, without imposing the burden of ownership management that comes from using a `unique_ptr`. Note that if you really _do_ need the ownership management, than you can do a `std::unique_ptr<rmm::device_buffer> my_ptr`. 

See https://github.com/rapidsai/rmm/issues/33

Reply:
@jrhemstad I don't see what you mean by "burden of ownership management"
The implementation of rmm::device_buffer would need to handle the same use cases as unique_ptr, such as copy and move construction. So it will either converge towards unique_ptr behavior and implementation, or it would behave differently and surprise its users.

Reply:
@j-ieong I know you also had a preference on this. Can you please comment on the preferred approach here?

Reply:
@jrhemstad On the subject of ownership management: I expect to need the unique_ptr behavior because I plan to use this type to pass CUDA buffers as return results from functions. Without correct behavior on copy/move, this can cause a number of issues.

Reply:
I think @jrhemstad means that all our current usage only requires lifetime management so we can get away with a simple `rmm::device_buffer` wrapper.

If we do need ownership, then the suggestion of `std::unique_ptr<rmm::device_buffer>` will provide both while keeping things separate and not placing additional overhead on code that only requires lifetime management.

Additionally, a client who wants ownership may want refcounted buffers (i.e. `shared_ptr`) so only having `unique_ptr` may be too limiting...

Reply:
Using a `unique_ptr` prevents you from adding any additional functionality such as `resize` or `reserve`.

What I am proposing is something very similar to what was done in cuML with their [`device_buffer` class](https://github.com/rapidsai/cuml/blob/branch-0.7/ml-prims/src/common/device_buffer.hpp).

> On the subject of ownership management: I expect to need the unique_ptr behavior because I plan to use this type to pass CUDA buffers as return results from functions. Without correct behavior on copy/move, this can cause a number of issues.

Sure, and you can provide this via a `std::unique_ptr< rmm::device_buffer >`. This approach provides the _option_ to use the ownership semantics of `unique_ptr`, but doesn't require that it _always_ be a `unique_ptr`.

Reply:
@j-ieong makes a great point that you may sometimes want to use a `shared_ptr`.

The main point is, I believe `rmm::device_buffer` provides more generic flexibility on top of which you can do things like use a `unique_ptr` or a `shared_ptr`.

Reply:
I see, decouple ownership from behavior on move/copy.
In this case, the device_buffer would be a container that copies the cuda buffer on copy, right?

@harrism any concerns about this direction change?

Reply:
Yeah, I think "burden of ownership management" needs better explanation.  "[std::unique_ptr is a smart pointer that owns and manages another object through a pointer and disposes of that object when the unique_ptr goes out of scope.](https://en.cppreference.com/w/cpp/memory/unique_ptr)" Doesn't sound too burdensome and fits the use case here, doesn't it?

Reply:
OK, so we need an rmm::device_buffer. I would prefer to have that done in the RMM repo, not here. So how do we move this PR forward?

Reply:
@harrism I can include a rmm_device_buffer class in this PR, and it can be easily moved to rmm repo in a separate PR.